src/main/java/com/google/cloud/tools/managedcloudsdk/internal/extract/TarGzExtractor.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.managedcloudsdk.internal.extract;
+
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.PosixFileAttributeView;
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
+import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream;
+import org.apache.commons.compress.utils.IOUtils;
+
+/**
+ * Extractor for *.tar.gz google cloud sdk archives.
+ *
+ * <p>NOTE: this does not handle links or symlinks or any other kind of special types in the tar. It
+ * will only create files and directories.
+ */
+public class TarGzExtractor implements Extractor {
+
+  private final Path archive;
+  private final Path destination;
+  private final ExtractorMessageListener listener;
+
+  TarGzExtractor(Path archive, Path destination, ExtractorMessageListener listener) {
+    this.archive = archive;
+    this.destination = destination;
+    this.listener = listener;
+  }
+
+  @Override
+  public Path call() throws IOException {
+    listener.message("Extracting archive: " + archive.toString());
+
+    GzipCompressorInputStream gzipIn = new GzipCompressorInputStream(Files.newInputStream(archive));
+    try (TarArchiveInputStream in = new TarArchiveInputStream(gzipIn)) {
+      TarArchiveEntry entry;
+      while ((entry = in.getNextTarEntry()) != null) {
+        final Path entryPath = destination.resolve(entry.getName());
+        if (listener != null) {
+          listener.message(entryPath.toString());
+        }
+        if (entry.isDirectory()) {
+          if (!Files.exists(entryPath)) {
```

Comment:
Why is this checking if the `entryPath` doesn't exist? Shouldn't that never happen, and if it did happen, wouldn't it mean something went wrong?

Reply:
I think directories can be out of order.
For example, the following situation where a sub dir is created before a parent, but all entries are in the tar :
```
$ tree .
.
└── a
    ├── b
    │   └── c.txt
    └── d.txt

$ tar -cf test.tar a/b/ a/

$ tar -tf test.tar 
a/b/
a/b/c.txt
a/
a/b/
a/b/c.txt
a/d.txt
```

Reply:
Hmm in that case, if a file came before its directory, like this:

```
$ tar -cf test.tar a/b/c.txt a/b/ a/
$ tar -tf test.tar 
a/b/c.txt
a/b/
...
```

Then the code breaks with a NoSuchFileException.

Reply:
ugh, stupid tar, okay yeah I'll handle that.